pkg/dnscache/nodednscache.go

```diff
@@ -0,0 +1,139 @@
+package dnscache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+
+	"github.com/aquasecurity/tracee/types/detect"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+type NodeCache struct {
+	queryRoots   *expirable.LRU[string, *dnsQueryNode]
+	queryIndices map[string]*dnsQueryNode
+
+	lock *sync.RWMutex
+}
+
+func NewNodeDNSCache(config Config) *NodeCache {
+	ttl := time.Second * time.Duration(config.TTL)
+	nc := &NodeCache{}
+	nc.queryRoots = expirable.NewLRU[string, *dnsQueryNode](config.CacheSize, nc.evictQueryRoot, ttl)
+	nc.queryIndices = make(map[string]*dnsQueryNode)
+	nc.lock = new(sync.RWMutex)
+	return nc
+}
+
+func (nc *NodeCache) Add(dns *trace.ProtoDNS) error {
+	// discover if it is a request or response
+
+	if dns.QR != 1 {
+		return nil // not a dns response
+	}
+
+	if len(dns.Questions) != 1 {
+		return fmt.Errorf("wrong number of requests found")
+	}
+
+	nc.lock.Lock()
+	defer nc.lock.Unlock()
+
+	question := dns.Questions[0].Name
+	questionNode, ok := nc.queryIndices[question]
+	// Check if question is indexed in the tree...
+	if !ok {
+		nc.addRootNode(dns)
+	} else {
+		nc.addChildNodes(dns, questionNode)
+	}
+	return nil
+}
+
+// Get returns all parent and child DNS records relative to the given record value.
+// Note, that the query does not traverse the tree downwards from the parent nodes.
+func (nc *NodeCache) Get(key string) (nodeCacheQuery, error) {
+	nc.lock.RLock()
+	defer nc.lock.RUnlock()
+
+	node, ok := nc.queryIndices[key]
+	if !ok {
+		return nodeCacheQuery{}, detect.ErrDataNotFound
+	}
+
+	queryResult := nodeCacheQuery{
+		dnsResults: []string{},
+		ipResults:  []string{},
+	}
+
+	err := addNodeToQueryResult(node, &queryResult, false)
+	if err != nil {
+		return queryResult, err
+	}
+
+	for node != nil {
+		node = node.previous
+		if node != nil {
+			err = addSingleNodeToQueryResult(node, &queryResult, true)
+			if err != nil {
+				return queryResult, err
+			}
+		}
+	}
+
+	return queryResult, nil
+}
+
+// addRootNode adds a new root node for DNS queries
+func (nc *NodeCache) addRootNode(dns *trace.ProtoDNS) {
+	nodeType := DNS
+	if dns.Questions[0].Type == "PTR" {
+		// Reverse Query
+		nodeType = IP
+	}
+
+	node := &dnsQueryNode{
+		value:    dns.Questions[0].Name,
+		nodeType: nodeType,
+		next:     make([]*dnsQueryNode, 0, len(dns.Answers)),
+	}
+
+	nc.addChildNodes(dns, node)
+	nc.queryRoots.Add(node.value, node)
+	nc.queryIndices[node.value] = node
+}
+
+// addChildNodes builds new DNS nodes from a ProtoDNS object and attaches them to a parent node
+func (nc *NodeCache) addChildNodes(dns *trace.ProtoDNS, parent *dnsQueryNode) {
+	for _, answer := range dns.Answers {
+		node := makeNodeFromAnswer(parent, &answer)
+		if node != nil {
+			parent.next = append(parent.next, node)
+			nc.queryIndices[node.value] = node
+		}
```

Comment:
In here, consider two different DNS queries to the same IP as an answer. Using the IP address as the index (since you're now creating a node from the answer) will cause the second query's answer to overwrite the first in nc.queryIndices, no ?

Reply:
Hmmm, so we can do two options here:
1. make queryIndices a map of value to a list of valid nodes, and keep the graph as a tree
2. give up on a tree structure and have the option for multiple parents (which I think would basically make it just a graph).

I slightly prefer option 2.

Reply:
I believe we should not pick PTR resolutions now. Mainly because a reverse IP could be something totally different than any other query to that name. I think you don't have to be authoritative to what you respond for PTR resolutions.

About having a graph, in order to extract Names or IPs from any data structure we might have, I believe we could have a Bidirectional Graph (it would have duplicates when compared to DAGs BUT it would allow going back and forth from any point).

![image](https://github.com/aquasecurity/tracee/assets/7395852/3abf9f28-0d9f-460e-9adc-94c1ef675666)

This example is simple and looks like a double linked list, but if we make things more complicated:

![image](https://github.com/aquasecurity/tracee/assets/7395852/b058cdf3-8296-43b6-a8c3-4e0fc8bdc513)

Also, I would suggest (even require) that you use a good/known/supported package to create such data structures (https://github.com/gonum/gonum/tree/master for example).

Reply:
BTW, for the record, we gave up on this idea as @NDStrahilevitz saw things would get more complicated and documenting/re-structuring his initial data structures were enough.